File: swift/2807_insert_greatest_common_divisors.swift
public class ListNode {
    public var val: Int
    public var next: ListNode?
    public init() { val = 0; next = nil }
    public init(_ val: Int) { self.val = val; next = nil }
    public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next }
}

class Solution {
    func insertGreatestCommonDivisors(_ head: ListNode?) -> ListNode? {
        var currentNode = head

        while let current = currentNode, let next = current.next {
            let newLinkedList = ListNode(gcd(current.val, next.val))

            newLinkedList.next = next
            current.next = newLinkedList
            currentNode = newLinkedList.next
        }

        return head
    }

    func gcd(_ a: Int, _ b: Int) -> Int {
        return b == 0 ? a : gcd(b, a % b)
    }
}

let h = ListNode(18, ListNode(6, ListNode(10, ListNode(3))))
let newLinkedList = Solution().insertGreatestCommonDivisors(h)
if let newLinkedList = newLinkedList {
    print("\(newLinkedList)")
} else {
    print("newLinkedList is nil")
}

File: ts/src/leetcode/2807_insert_greatest_common_divisors.ts
import { log } from "console";

export class ListNode<T> {
  val: T | null;
  next: ListNode<T> | null;
  constructor(val?: T, next?: ListNode<T> | null) {
    this.val = val === undefined ? null : val;
    this.next = next === undefined ? null : next;
  }

  forEach(callback: (node: T | null) => void) {
    let currentNode: ListNode<T> | null = this;

    while (currentNode) {
      callback(currentNode?.val);
      currentNode = currentNode.next;
    }
  }
}

function gcd(n: number, m: number): number {
  while (m !== 0) {
    let temp = m;
    m = n % m;
    n = temp;
  }

  return n;
}

function insertGreatestCommonDivisors(
  head: ListNode<number> | null,
): ListNode<number> | null {
  let currentNode = head;

  while (
    currentNode?.val !== null &&
    currentNode?.next &&
    currentNode.next?.val !== null
  ) {
    const newNode = new ListNode(gcd(currentNode.val, currentNode.next.val));
    log(newNode);
    newNode.next = currentNode.next;
    currentNode.next = newNode;
    currentNode = newNode.next;
  }

  return head;
}

export default insertGreatestCommonDivisors;
